erdos/ray/ray_operator.py

```diff
@@ -52,15 +52,46 @@ def on_completion_msg(self, msg):
         """Invokes corresponding callback for stream stream_name."""
         self._op.log_event(time.time(), msg.timestamp,
                            'receive watermark {}'.format(msg.stream_name))
-        for cb in self._completion_callbacks.get(msg.stream_uid, []):
-            cb(msg)
+
+        # Ensure that the watermark is monotonically increasing.
+        high_watermark = self._op._stream_to_high_watermark[msg.stream_name]
+        if not high_watermark:
+            # The first watermark, just set the dictionary with the value.
+            self._op._stream_to_high_watermark[
+                msg.stream_name] = msg.timestamp
+        else:
+            if high_watermark >= msg.timestamp:
+                raise Exception(
+                    "The watermark received in the msg {} is not "
+                    "higher than the watermark previously received "
+                    "on the same stream: {}".format(msg, high_watermark))
+            else:
+                self._op._stream_to_high_watermark[msg.stream_name] = \
+                        msg.timestamp
+
+        # Now check if all other streams have a higher or equal watermark.
+        # If yes, flow this watermark. If not, return from this function
+        # Also, maintain the lowest watermark observed.
+        low_watermark = msg.timestamp
+        for stream, watermark in self._op._stream_to_high_watermark.items():
+            if stream != msg.stream_name:
+                if not watermark or watermark < msg.timestamp:
+                    return
+                if low_watermark > watermark:
```

Comment:
Doesn't this work correctly when you have two input streams (i1 and i2), i2 has received watermark t + 1, and i1 just received watermark t? If I understand the code correctly, it will update low_watermark to t +1, and send watermark t + 1, instead of t.

Reply:
I take it back :)